File: containers/app4/test/go/database_example.go
package main

import (
	"database/sql"
	"fmt"
	"log"
	"os"

	_ "github.com/mattn/go-sqlite3"
)

// User represents a user in the database
type User struct {
	ID    int
	Name  string
	Email string
}

// Database represents the database connection
type Database struct {
	db *sql.DB
}

// NewDatabase creates a new database connection
func NewDatabase() (*Database, error) {
	// Remove existing database file if it exists
	os.Remove("users.db")

	// Open database connection
	db, err := sql.Open("sqlite3", "users.db")
	if err != nil {
		return nil, err
	}

	// Create users table
	createTableSQL := `
	CREATE TABLE users (
		id INTEGER PRIMARY KEY AUTOINCREMENT,
		name TEXT NOT NULL,
		email TEXT NOT NULL UNIQUE
	)`

	_, err = db.Exec(createTableSQL)
	if err != nil {
		return nil, err
	}

	fmt.Println("Database created successfully")
	fmt.Println("Table created successfully")

	return &Database{db: db}, nil
}

// InsertUser inserts a new user into the database
func (d *Database) InsertUser(name, email string) error {
	insertSQL := "INSERT INTO users (name, email) VALUES (?, ?)"
	_, err := d.db.Exec(insertSQL, name, email)
	if err != nil {
		return err
	}
	fmt.Println("User inserted successfully")
	return nil
}

// GetAllUsers retrieves all users from the database
func (d *Database) GetAllUsers() ([]User, error) {
	rows, err := d.db.Query("SELECT id, name, email FROM users")
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var users []User
	for rows.Next() {
		var user User
		err := rows.Scan(&user.ID, &user.Name, &user.Email)
		if err != nil {
			return nil, err
		}
		users = append(users, user)
	}
	return users, nil
}

// DisplayUsers displays all users in a formatted way
func DisplayUsers(users []User) {
	fmt.Println("All users:")
	for _, user := range users {
		fmt.Printf("ID: %d, Name: %s, Email: %s\n", user.ID, user.Name, user.Email)
	}
}

// Close closes the database connection
func (d *Database) Close() error {
	return d.db.Close()
}

func main() {
	fmt.Println("Go SQLite Database Example")
	fmt.Println("==========================")

	// Create database
	database, err := NewDatabase()
	if err != nil {
		log.Fatal("Failed to create database:", err)
	}
	defer database.Close()

	// Insert sample users
	err = database.InsertUser("John Doe", "john@example.com")
	if err != nil {
		log.Printf("Failed to insert user: %v", err)
	}

	err = database.InsertUser("Jane Smith", "jane@example.com")
	if err != nil {
		log.Printf("Failed to insert user: %v", err)
	}

	err = database.InsertUser("Bob Johnson", "bob@example.com")
	if err != nil {
		log.Printf("Failed to insert user: %v", err)
	}

	// Query and display all users
	users, err := database.GetAllUsers()
	if err != nil {
		log.Printf("Failed to get users: %v", err)
	} else {
		DisplayUsers(users)
	}

	// Close database connection
	database.Close()
	fmt.Println("Database connection closed")
}

File: containers/python/sample-project/database_example.py
#!/usr/bin/env python3
"""
Simple Database Example
A basic Python script demonstrating SQLite database operations.
"""

import sqlite3
import os

def create_database():
    """Create database and table."""
    conn = sqlite3.connect('users.db')
    cursor = conn.cursor()
    
    # Create users table
    cursor.execute('''
        CREATE TABLE IF NOT EXISTS users (
            id INTEGER PRIMARY KEY,
            name TEXT NOT NULL,
            email TEXT NOT NULL
        )
    ''')
    
    conn.commit()
    conn.close()
    print("✅ Database created successfully!")

def add_user(name, email):
    """Add a new user to the database."""
    conn = sqlite3.connect('users.db')
    cursor = conn.cursor()
    
    cursor.execute('INSERT INTO users (name, email) VALUES (?, ?)', (name, email))
    conn.commit()
    conn.close()
    print(f"✅ User '{name}' added successfully!")

def view_all_users():
    """View all users in the database."""
    conn = sqlite3.connect('users.db')
    cursor = conn.cursor()
    
    cursor.execute('SELECT * FROM users')
    users = cursor.fetchall()
    
    if users:
        print("\n📋 All Users:")
        print("-" * 40)
        for user in users:
            print(f"ID: {user[0]}, Name: {user[1]}, Email: {user[2]}")
    else:
        print("📭 No users found in database.")
    
    conn.close()

def delete_user(user_id):
    """Delete a user by ID."""
    conn = sqlite3.connect('users.db')
    cursor = conn.cursor()
    
    cursor.execute('DELETE FROM users WHERE id = ?', (user_id,))
    conn.commit()
    conn.close()
    print(f"✅ User with ID {user_id} deleted!")

def main():
    """Main function to run the database example."""
    print("🗄️  Simple Database Example")
    print("=" * 30)
    
    # Create database
    create_database()
    
    # Add some sample users
    print("\n📝 Adding sample users...")
    add_user("John Doe", "john@example.com")
    add_user("Jane Smith", "jane@example.com")
    add_user("Bob Johnson", "bob@example.com")
    
    # View all users
    view_all_users()
    
    # Interactive menu
    while True:
        print("\n🔧 What would you like to do?")
        print("1. Add a new user")
        print("2. View all users")
        print("3. Delete a user")
        print("4. Exit")
        
        choice = input("\nEnter your choice (1-4): ").strip()
        
        if choice == '1':
            name = input("Enter name: ").strip()
            email = input("Enter email: ").strip()
            if name and email:
                add_user(name, email)
            else:
                print("❌ Name and email are required!")
        
        elif choice == '2':
            view_all_users()
        
        elif choice == '3':
            view_all_users()
            try:
                user_id = int(input("Enter user ID to delete: ").strip())
                delete_user(user_id)
            except ValueError:
                print("❌ Please enter a valid user ID!")
        
        elif choice == '4':
            print("👋 Goodbye!")
            break
        
        else:
            print("❌ Invalid choice! Please enter 1-4.")

if __name__ == "__main__":
    main()
